File: src/main/java/schemabuilder/processor/pipelines/building/WiringBuilder.java
package schemabuilder.processor.pipelines.building;

import graphql.schema.idl.RuntimeWiring;
import graphql.schema.idl.TypeRuntimeWiring;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import schemabuilder.processor.pipelines.parsing.GraphQLClassParser;
import schemabuilder.processor.pipelines.parsing.datafetchers.GraphQLDataFetcherBank;
import schemabuilder.processor.pipelines.parsing.datafetchers.GraphQLDataFetcherType;
import schemabuilder.processor.pipelines.parsing.directives.GraphQLDirectiveBank;
import schemabuilder.processor.pipelines.parsing.directives.GraphQLDirectiveType;
import schemabuilder.processor.pipelines.parsing.scalars.GraphQLScalarBank;
import schemabuilder.processor.pipelines.parsing.scalars.GraphQLScalarType;
import schemabuilder.processor.pipelines.parsing.typeresolvers.GraphQLTypeResolverBank;
import schemabuilder.processor.pipelines.parsing.typeresolvers.GraphQLTypeResolverType;

public class WiringBuilder {

    private GraphQLDataFetcherBank dataFetchers = GraphQLDataFetcherBank.getInstance();
    private GraphQLDirectiveBank directives = GraphQLDirectiveBank.getInstance();
    private GraphQLScalarBank scalars = GraphQLScalarBank.getInstance();
    private GraphQLTypeResolverBank typeResolvers = GraphQLTypeResolverBank.getInstance();


    public static WiringBuilder withOptions(String basePackage, Set<Class<?>> clazzes) {
        return new WiringBuilder(basePackage, clazzes);
    }

    private WiringBuilder(String basePackage, Set<Class<?>> clazzes) {
        new GraphQLClassParser(basePackage, clazzes).parseClasses();
    }


    public RuntimeWiring.Builder buildWiring() {
        Map<String, Set<GraphQLDataFetcherType>> typeMap = new HashMap<>();

        for (GraphQLDataFetcherType dataFetcher : this.dataFetchers.getDataFetchers()) {
            typeMap.computeIfAbsent(dataFetcher.getTypeName(), k -> new HashSet<>());
            typeMap.get(dataFetcher.getTypeName()).add(dataFetcher);
        }


        RuntimeWiring.Builder builder = RuntimeWiring.newRuntimeWiring();

        for (String typeName : typeMap.keySet()) {
            TypeRuntimeWiring.Builder typeBuilder = TypeRuntimeWiring.newTypeWiring(typeName);

            for (GraphQLDataFetcherType dataFetcher : this.dataFetchers.getDataFetchers()) {
                typeBuilder.dataFetcher(dataFetcher.getName(), dataFetcher.getDataFetcher());
            }

            builder.type(typeBuilder);
        }

        for (GraphQLTypeResolverType resolver : this.typeResolvers.getTypeResolvers()) {
            TypeRuntimeWiring.Builder typeResolverBuilder = TypeRuntimeWiring.newTypeWiring(resolver.getName());
            typeResolverBuilder.typeResolver(resolver.getResolver());
            builder.type(typeResolverBuilder);
        }


        for (GraphQLScalarType scalar : this.scalars.getScalars()) {
            builder.scalar(scalar.getScalar());
        }

        for (GraphQLDirectiveType directive : this.directives.getDirectives()) {
            builder.directive(directive.getName(), directive.getDirective());
        }

        return builder;
    }
}

File: src/main/java/schemabuilder/processor/pipelines/building/WiringBuilder.kt
package schemabuilder.processor.pipelines.building

import graphql.schema.idl.RuntimeWiring
import graphql.schema.idl.TypeRuntimeWiring
import schemabuilder.processor.pipelines.parsing.GraphQLClassParser
import schemabuilder.processor.pipelines.parsing.ParsedResults
import schemabuilder.processor.pipelines.parsing.datafetchers.GraphQLDataFetcherType
import schemabuilder.processor.wiring.InstanceFetcher
import java.util.*

class WiringBuilder private constructor(basePackage: String?, clazzes: Set<Class<*>>, fetcher: InstanceFetcher) {

    fun buildWiring(): RuntimeWiring.Builder {
        val typeMap: MutableMap<String?, MutableSet<GraphQLDataFetcherType>> = HashMap()
        for (dataFetcher in ParsedResults.datafetchers) {
            typeMap.computeIfAbsent(dataFetcher.typeName) { HashSet() }
            typeMap[dataFetcher.typeName]?.add(dataFetcher)
        }

        val builder = RuntimeWiring.newRuntimeWiring()
        for (typeName in typeMap.keys) {
            val typeBuilder = TypeRuntimeWiring.newTypeWiring(typeName)
            println(typeName)
            for (dataFetcher in typeMap[typeName]!!) {
                typeBuilder.dataFetcher(dataFetcher.name, dataFetcher.dataFetcher)
                println("-" + dataFetcher.name)
            }
            builder.type(typeBuilder)
        }

        for (resolver in ParsedResults.typeResolvers) {
            val typeResolverBuilder = TypeRuntimeWiring.newTypeWiring(resolver.name)
            typeResolverBuilder.typeResolver(resolver.resolver)
            builder.type(typeResolverBuilder)
        }

        for (scalar in ParsedResults.scalars) {
            builder.scalar(scalar.scalar)
        }

        for (directive in ParsedResults.directives) {
            builder.directive(directive.name, directive.directive)
        }
        return builder
    }

    companion object {
        fun withOptions(basePackage: String?, clazzes: Set<Class<*>>, fetcher: InstanceFetcher): WiringBuilder {
            return WiringBuilder(basePackage, clazzes, fetcher)
        }
    }

    init {
        GraphQLClassParser(basePackage, clazzes, fetcher).parseClasses()
    }
}